src/network/sntp_client.c

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2023 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/logging/log.h>
+#include "sntp_client.h"
+#include "netmon_stats.h"
+
+LOG_MODULE_REGISTER(sntp_client, CONFIG_LOG_DEFAULT_LEVEL);
+
+/* Store elapsed real time since the system booted in milliseconds. */
+static int64_t time_base_ms;
+
+/* Get the timestamp in seconds since 1st Jan 1970 and store it
+ * in the time_base tracking variable.
+ */
+int sntp_client_get_time(time_t *sec)
+{
+	int ret;
+	struct sntp_time sntp_time;
+	// struct timespec time_spec;
+
+	LOG_INF("Sending SNTP request for current time:");
+
+	ret = sntp_simple(CONFIG_SNTP_INIT_SERVER, SYS_FOREVER_MS, &sntp_time);
+	if (ret < 0) {
+		LOG_ERR("Failed to SNTP current time: %d", ret);
+		return ret;
+	}
+
+	*sec = sntp_time.seconds;
+
+	// /* TODO:  Update the system clock time */
+	// time_spec.tv_sec = sntp_time.seconds;
+	// time_spec.tv_nsec = ((uint64_t)sntp_time.fraction * (1000 * 1000 * 1000)) >> 32;
+	// ret = clock_settime(CLOCK_REALTIME, &time_spec);
+	// if (ret != 0) {
+	// 	LOG_ERR("Failed to set system time, %d", ret);
+	// }
+
+    /* Increment the sntp time req stats counter */
+    netmon_sntp_time_req_count_stats(get_netmon_stats_inst());
+
+	return ret;
+}
+
+/* Get the current time in seconds since 1st Jan 1970.
+ */
+time_t get_current_time()
+{
+	time_t current_time_sec;
+
+	current_time_sec = (time_t)((k_uptime_get() + time_base_ms) / 1000);
+
+	return current_time_sec;
+}
+
+/* Get the real date and timestamp info in the time management structure. */
+void get_current_date_time(struct tm *info)
+{
+	time_t current_time_sec;
+
+	current_time_sec = get_current_time();
+	gmtime_r(&current_time_sec, info);
+	LOG_INF("Data and time: %04u-%02u-%02u %02u:%02u:%02u",
+		info->tm_year + 1900, info->tm_mon + 1, info->tm_mday,
+		info->tm_hour, info->tm_min, info->tm_sec);
+}
+
+/* Get the latest time from the SNTP server and align the time base. */
+void sntp_resync_time()
+{
+	int ret;
+	time_t sec;
+	struct tm info;
+
+	ret = sntp_client_get_time(&sec);
+	if (ret < 0) {
+		LOG_ERR("Failed to sync time: %d", ret);
+		return;
+	}
+	time_base_ms = (sec * MSEC_PER_SEC - k_uptime_get());
+	get_current_date_time(&info);
+}
+
+void sntp_client_init()
+{
```

Comment:
As I understand, this API maintains it's own time base. Would we be better off working with the existing and posix time functions, which already have a mechanism for dealing with a base.

Regardless of this, are there any concerns about time skips that come about because the system clock runs at a different speed than the real world time? It is possible to get either a discontinuity, or for time to appear to go backward.

Also, how does `k_uptime_get` respond to systems that enter low power mode?